<launch>

  <!-- DEFAULT PARAMETERS AND FRAMES -->
  <arg name="uav_name" />
  <arg name="fcu_frame" />
  <arg name="lidar_frame" />
  <arg name="map_frame" />

    <!-- SENSOR PARAMETERS -->
  <arg name="vertical_fov" />
  <arg name="frequency" />

    <!-- FILTRATION PARAMETERS -->
  <arg name="custom_preprocessing" />
  <arg name="min_range" />
  <arg name="max_range" />
  
  <arg name="nodelet" default="standalone" />
  <arg name="nodelet_manager" default="" />
  <arg name="launch_prefix" default="" />

  <group ns="$(arg uav_name)">

    <!-- Odometry -->
    <!-- <node pkg="nodelet" type="nodelet" name="aloam_odometry" args="$(arg nodelet) aloam_velodyne/aloamOdometry $(arg nodelet_manager)" output="screen" launch-prefix=""> -->
    <node pkg="aloam_velodyne" type="aloamOdometry" name="aloam_odometry" output="screen">
      <param name="debug" type="bool" value="$(optenv DEBUG false)" />
      <param name="scan_line" type="int" value="16" />

      <param name="uav_name" type="string" value="$(arg uav_name)"/>
      <param name="map_frame" type="string" value="$(arg map_frame)"/>
      <param name="lidar_frame" type="string" value="$(arg lidar_frame)"/>
      <param name="fcu_frame" type="string" value="$(arg fcu_frame)"/>

        <!-- if 1, do mapping 10 Hz, if 2, do mapping 5 Hz. Suggest to use 1, it will adjust frequence automaticlly -->
      <param name="mapping_skip_frame" type="int" value="1" />

      <param name="perform_scan_preprocessing" type="bool" value="$(arg custom_preprocessing)" />
      <param name="min_range" type="double" value="$(arg min_range)"/>
      <param name="max_range" type="double" value="$(arg max_range)"/>
      <param name="vertical_fov" type="double" value="$(arg vertical_fov)"/>
      <param name="frequency" type="double" value="$(arg frequency)"/>

      <!-- Subscribers -->
      <remap if="$(arg custom_preprocessing)" from="/sensor_points" to="os1_cloud_node/points_processed"/>
      <remap unless="$(arg custom_preprocessing)" from="/sensor_points" to="os1_cloud_node/points"/>

      <remap from="/mavros_odom_in" to="mavros_interface/converted_mavros_odom"/>
      <remap from="/main_odom_in" to="odometry/odom_main"/>

      <!-- Publishers -->
      <remap from="/laser_odom_to_init" to="~laser_odom_to_init"/>

      <remap from="/aft_mapped_path" to="aloam_mapping/aft_mapped_path"/> 
      <remap from="/aft_mapped_to_init" to="aloam_mapping/aft_mapped_to_init"/> 
      <remap from="/laser_cloud_map" to="aloam_mapping/laser_cloud_map"/> 
      <remap from="/sensor_cloud_registered" to="aloam_mapping/sensor_cloud_registered"/> 
      
    </node>
  </group>

</launch>
